### DISCUSSION TITLE
JavascriptObjectRepository seems can not effect UI thread?

### DISCUSSION THREAD

**Discussion by tp1415926535** [cefsharp/CefSharp#3768] [Q&A] [answered]
I followed the "[How do you expose a .NET class to JavaScript](https://github.com/cefsharp/CefSharp/wiki/General-Usage#3-how-do-you-expose-a-net-class-to-javascript)" and achieved it, but when I change UI in class method, there will be some error in console when call it, like this:
C# code:
> 
        browser.JavascriptObjectRepository.Register("boundAsync", new BoundObject(this), true, BindingOptions.DefaultBinder);

        public class BoundObject
        { 
            private MainWindow mainWindow;
            public BoundObject(MainWindow MainWindow)
            {
                mainWindow = MainWindow;
            }
            public int Add(int a, int b)
            {                
                mainWindow.Title = "new Title";
                return a + b;
            }
        }

JS code:
> 
        (async function()
        {
	        await CefSharp.BindObjectAsync("boundAsync");
	        boundAsync.add(16, 2);
        })();

error:
> InvalidOperationException"(located in WindowsBase.dll)
TargetInvocationException"(located in mscorlib.dll)
InvalidOperationException"(located in CefSharp.dll)
[0822/003331.413:INFO:CONSOLE(0)] "Uncaught (in promise) System.InvalidOperationException: Could not execute method: add(16, 2)  ---> System.Reflection.TargetInvocationException:XXXXXXX---> System.InvalidOperationException: XXXXX
System.Windows.Threading.Dispatcher.VerifyAccess()
System.Windows.Window.set_Title(String value)
....

so is there a way like *webbrowser.ObjectForScripting* can effect to the window UI ?

**Comment by ukandrewc** [comment, marked as answer]
Assuming boundObject was created on your UI thread, you need to invoke mainWindow.Title because the call into boundObject.add is made on the CEF UI thread.

PS: The stacktrace tells you that because it failed on `Dispatcher.VerifyAccess()`, which checks that you are calling from the same thread.

**Reply by amaitland** [reply]
> because the call into boundObject.add is made on the CEF UI thread.

Methods are actually executed on the thread pool. 

> so is there a way like webbrowser.ObjectForScripting can effect to the window UI ?

@tp1415926535 you can use the dispatcher to access the UI thread. Your MainWindow will have a Dispatcher property. 

https://docs.microsoft.com/en-us/dotnet/api/system.windows.threading.dispatcher.invokeasync?view=net-5.0#System_Windows_Threading_Dispatcher_InvokeAsync_System_Action_

**Reply by tp1415926535** [reply]
> @tp1415926535 you can use the dispatcher to access the UI thread. Your MainWindow will have a Dispatcher property.
> 
> https://docs.microsoft.com/en-us/dotnet/api/system.windows.threading.dispatcher.invokeasync?view=net-5.0#System_Windows_Threading_Dispatcher_InvokeAsync_System_Action_

well... I make it after add a Invoke method there :

> 
            public int Add(int a, int b)
            {
                mainWindow.Dispatcher.Invoke(new Action(() =>
                {
                    mainWindow.Title = "new Title";
                }));
                return a + b;
            }

it works, thanks.

**Reply by amaitland** [reply]
Use InvokeAsync instead of Invoke.

**Reply by ukandrewc** [reply]
> Methods are actually executed on the thread pool.

Aha, worth knowing ;-)

**Reply by tp1415926535** [reply]
> 使用 InvokeAsync 而不是 Invoke。

Oh, good idea.
